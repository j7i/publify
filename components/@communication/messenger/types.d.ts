import { IMessage } from '@communication/chat/chatController/types'

export interface IMessengerProps {
  user: firebase.User
}

export interface IMessengerState {
  currentChat?: string
  chats: IChat[]
}

export interface IChat {
  id: string
  messages: IMessage[]
  members: string[]
  advertId: string
}
