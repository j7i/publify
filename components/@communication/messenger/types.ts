import { IMessage } from '@communication/chat/types'
import { ReactNode } from 'react'

export interface IMessengerProps {
  user: firebase.User
}

export interface IMessengerState {
  currentChat?: string
  chats: IChat[]
  notification: ReactNode
}

export interface IChat {
  id: string
  messages: IMessage[]
  members: string[]
  memberInfos: IMemberInfos
  advertId: string
  advertTitle: string
}

interface IMemberInfos {
  [key: string]: IMemberInfo
}

interface IMemberInfo {
  name: string
  userImageURL: string
}
