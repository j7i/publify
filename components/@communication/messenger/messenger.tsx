import { Chat } from '@communication/chat'
import { FirebaseCollection } from '@config'
import { List, ListItem, ListItemIcon, ListItemText, ListSubheader, Paper } from '@material-ui/core'
import ChatIcon from '@material-ui/icons/Chat'
import firebase from 'firebase'
import { PureComponent, ReactNode } from 'react'
import styles from './messengerStyles.css'
import { IChat, IMessengerProps, IMessengerState } from './types'

export class Messenger extends PureComponent<IMessengerProps, IMessengerState> {
  public state: IMessengerState = {
    chats: [],
    currentChat: ''
  }

  public componentDidMount(): void {
    this.getUserSpecificChats()
  }

  public render(): ReactNode {
    const { chats, currentChat } = this.state

    return (
      <Paper className={styles.messenger}>
        <List className={styles.chatList} component="nav" subheader={<ListSubheader>Messages</ListSubheader>}>
          {chats.length !== 0 &&
            chats.map((chat: IChat, index: number) => {
              return (
                <ListItem
                  className={styles.chatListItem}
                  key={index}
                  selected={chat.id === currentChat}
                  button
                  onClick={(): void => this.startConversation(chat.id)}
                >
                  <ListItemIcon>
                    <ChatIcon />
                  </ListItemIcon>
                  <ListItemText primary={chat.advertId} />
                </ListItem>
              )
            })}
        </List>
        {currentChat && <Chat chatId={currentChat} key={currentChat} displayMode="embedded" />}
      </Paper>
    )
  }

  private startConversation = (currentChat: string): void => {
    this.setState({
      currentChat
    })
  }

  private getUserSpecificChats = async (): Promise<void> => {
    const firestore = firebase.firestore()
    const { user } = this.props

    // tslint:disable-next-line:no-any
    let chats: any[] = []
    firestore
      .collection(FirebaseCollection.CHATS)
      .where('members', 'array-contains', user!.uid)
      .get()
      .then((querySnapshot: firebase.firestore.QuerySnapshot) => {
        querySnapshot.forEach((doc: firebase.firestore.QueryDocumentSnapshot) => {
          chats.push({ id: doc.id, ...doc.data() })
        })
      })
      .then(() => {
        this.setState({ chats })
      })
      .catch((error: Error) => {
        // tslint:disable-next-line:no-console
        console.error('Error adding document: ', error)
      })
  }
}
