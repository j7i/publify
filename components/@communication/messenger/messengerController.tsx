import { FirebaseCollection, firestore } from '@config'
import { NotificationSeverity, SnackbarNotification } from '@core'
import { PureComponent, ReactNode } from 'react'
import { IMessengerControllerProps, IMessengerControllerState, IMessengerViewRenderProps } from './types'

export class MessengerController extends PureComponent<IMessengerControllerProps, IMessengerControllerState> {
  public state: IMessengerControllerState = {
    chats: [],
    currentChat: '',
    notification: null
  }

  public componentDidMount(): void {
    this.getUserSpecificChats()
  }

  public render(): ReactNode {
    const { chats, currentChat } = this.state
    const { children, user } = this.props
    const renderProps: IMessengerViewRenderProps = {
      user,
      chats,
      currentChat,
      clearCurrentConversation: this.clearCurrentConversation,
      startConversation: this.startConversation
    }

    return children(renderProps)
  }

  private clearCurrentConversation = (): void => {
    this.setState({
      currentChat: ''
    })
  }

  private startConversation = (currentChat: string): void => {
    this.setState({
      currentChat
    })
  }

  private getUserSpecificChats = async (): Promise<void> => {
    // tslint:disable-next-line:no-any
    let chats: any[] = []
    const { user } = this.props
    firestore
      .collection(FirebaseCollection.CHATS)
      .where('members', 'array-contains', user!.uid)
      .get()
      .then((querySnapshot: firebase.firestore.QuerySnapshot) => {
        querySnapshot.forEach((doc: firebase.firestore.QueryDocumentSnapshot) => {
          chats.push({ id: doc.id, ...doc.data() })
        })
      })
      .then(() => {
        this.setState({ chats })
        if (window.innerWidth > 991) {
          this.setState({ currentChat: chats[0].id })
        }
      })
      .catch((error: Error) => {
        this.setState({
          notification: <SnackbarNotification key={Date.now() + Math.random()} message={error.message} severity={NotificationSeverity.ERROR} />
        })
      })
  }
}
