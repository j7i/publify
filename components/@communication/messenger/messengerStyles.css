.messenger {
  display: flex;
  flex-direction: row;
  width: 100%;
  height: calc(100vh - 64px);
  background-color: $grey1;
  /* padding: $spacingMedium; */
}

.chatList {
  margin-top: $spacingBase;
  display: flex;
  flex-basis: 1;
  flex-grow: 1;
  flex-direction: column;
  border-right: 1px solid $grey3;
  max-width: 400px;
  min-width: 300px;

  .chatListItem {
    &[class*='selected'] {
      background-color: $grey3;
    }
  }
}

.chatArea {
  flex-grow: 2;
  height: inherit;
}
