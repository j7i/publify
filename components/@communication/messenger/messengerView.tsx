import { Chat } from '@communication/chat'
import {
  AppBar,
  Avatar,
  Button,
  Divider,
  IconButton,
  List,
  ListItem,
  ListItemIcon,
  ListItemSecondaryAction,
  ListItemText,
  ListSubheader,
  Paper,
  Toolbar
} from '@material-ui/core'
import AccountCircle from '@material-ui/icons/AccountCircle'
import KeyboardBackspaceIcon from '@material-ui/icons/KeyboardBackspace'
import OpenInBrowserIcon from '@material-ui/icons/OpenInBrowser'
import classNames from 'classnames'
import Link from 'next/link'
import { Fragment, PureComponent, ReactNode } from 'react'
import styles from './messengerStyles.css'
import { IChat, IMessengerViewProps } from './types'

export class MessengerView extends PureComponent<IMessengerViewProps> {
  public render(): ReactNode {
    const { chats, currentChat, clearCurrentConversation } = this.props.renderProps

    return (
      <Paper className={styles.messenger}>
        <div className={classNames(styles.messengerAppBar, { [styles.active]: currentChat })}>
          <AppBar>
            <Toolbar>
              {currentChat ? (
                <>
                  <IconButton color="inherit" aria-label="Back" onClick={clearCurrentConversation}>
                    <KeyboardBackspaceIcon />
                  </IconButton>
                  Back
                  <div className={styles.grow} />
                </>
              ) : (
                <>Messages</>
              )}
            </Toolbar>
          </AppBar>
        </div>
        {chats.length !== 0 ? (
          <>
            {this.renderChatList()}
            {currentChat && (
              <div className={classNames(styles.messengerChatArea, styles.active)}>
                <Chat chatId={currentChat} key={currentChat} />
              </div>
            )}
          </>
        ) : (
          this.renderEmptyState()
        )}
      </Paper>
    )
  }

  private renderChatList = (): ReactNode => {
    const { currentChat, chats, user, startConversation } = this.props.renderProps

    return (
      <List
        className={classNames(styles.messengerChatList, { [styles.active]: !currentChat })}
        component="nav"
        subheader={
          <div>
            <ListSubheader className={styles.chatListSubheader}>Messages</ListSubheader>
            <Divider />
          </div>
        }
      >
        {chats.map((chat: IChat, index: number) => {
          const chatPartnerId = chat.members.filter((member: string) => member !== user.uid)
          const { name: chatPartnerFullName, userImageURL: chatPartnerImageURL } = chat.memberInfos[chatPartnerId[0]]
          const { id, advertTitle } = chat

          return (
            <Fragment key={index}>
              <ListItem className={styles.chatListItem} key={index} selected={id === currentChat} button onClick={(): void => startConversation(id)}>
                <ListItemIcon>{chatPartnerImageURL ? <Avatar alt={chatPartnerFullName} src={chatPartnerImageURL} /> : <AccountCircle />}</ListItemIcon>
                <ListItemText primary={chatPartnerFullName} secondary={advertTitle} />
                <ListItemSecondaryAction>
                  <Link href={`/detail/${chat.advertId}`}>
                    <IconButton aria-label="Delete">
                      <OpenInBrowserIcon />
                    </IconButton>
                  </Link>
                </ListItemSecondaryAction>
              </ListItem>
              <Divider />
            </Fragment>
          )
        })}
      </List>
    )
  }

  private renderEmptyState = (): ReactNode => (
    <div className={styles.emptyState}>
      <div className={styles.emptystateItems}>
        <h2>You actually have no active conversations.</h2>
        <Link href="/">
          <Button color="primary" variant="contained">
            Find an advert
          </Button>
        </Link>
      </div>
    </div>
  )
}
