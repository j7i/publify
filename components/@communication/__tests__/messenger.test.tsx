/* eslint-env jest */
// tslint:disable:no-object-literal-type-assertion
import { messengerChatsMock } from '@communication/__mocks__/messenger.mock'
import { MessengerView } from '@communication/messenger/messengerView'
import { cleanup, fireEvent, render } from 'react-testing-library'

jest.mock('@communication', () => 'Chat')
jest.mock('@core', () => 'GoogleMap')

describe('@communication/messenger:', () => {
  const clearCurrentConversation = jest.fn()
  const startConversation = jest.fn()
  const user = {} as firebase.User
  const withUser = { uid: '1a2b' } as firebase.User
  const renderPropsMock = {
    user,
    chats: messengerChatsMock,
    currentChat: '',
    clearCurrentConversation,
    startConversation
  }

  afterEach(cleanup)

  it('should render the messenger empty state', () => {
    // arrange
    const renderProps = { ...renderPropsMock, chats: [] }

    // act
    const { container } = render(<MessengerView renderProps={{ ...renderProps }} />)

    // assert
    expect(container).toMatchSnapshot('Representing the empty state without chats.')
  })

  it('should render a messenger list with two listItems', () => {
    const renderProps = { ...renderPropsMock, user: withUser }

    const { container } = render(<MessengerView renderProps={{ ...renderProps }} />)

    expect(container).toMatchSnapshot('Representing a basic messenger component with two availabel chats.')
  })

  it('should select a new chat as current', () => {
    const renderProps = { ...renderPropsMock, user: withUser }

    const { container } = render(<MessengerView renderProps={{ ...renderProps }} />)
    const listItem = container.querySelector('.chatListItem') as HTMLElement
    fireEvent.click(listItem)

    expect(startConversation).toHaveBeenCalled()
  })
})
