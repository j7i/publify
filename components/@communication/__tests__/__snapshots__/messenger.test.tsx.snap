// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@communication/messenger: should render a messenger list with two listItems: Representing a basic messenger component with two availabel chats. 1`] = `
<div>
  <div
    class="MuiPaper-root-77 MuiPaper-elevation2-81 MuiPaper-rounded-78 messenger"
  >
    <div
      class="messengerAppBar"
    >
      <header
        class="MuiPaper-root-77 MuiPaper-elevation4-83 MuiAppBar-root-104 MuiAppBar-positionFixed-105 MuiAppBar-colorPrimary-111 mui-fixed"
      >
        <div
          class="MuiToolbar-root-113 MuiToolbar-regular-115 MuiToolbar-gutters-114"
        >
          Messages
        </div>
      </header>
    </div>
    <nav
      class="MuiList-root-117 MuiList-padding-118 MuiList-subheader-120 messengerChatList active"
    >
      <div>
        <li
          class="MuiListSubheader-root-121 MuiListSubheader-sticky-126 MuiListSubheader-gutters-124 chatListSubheader"
        >
          Messages
        </li>
        <hr
          class="MuiDivider-root-127"
        />
      </div>
      <li
        class="MuiListItem-container-132"
      >
        <div
          class="MuiButtonBase-root-142 MuiListItem-root-131 MuiListItem-default-134 MuiListItem-gutters-138 MuiListItem-button-139 MuiListItem-secondaryAction-140 chatListItem"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root-145"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-146"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root-155"
          >
            <span
              class="MuiTypography-root-161 MuiTypography-subheading-168 MuiListItemText-primary-158"
            >
              Le
            </span>
            <p
              class="MuiTypography-root-161 MuiTypography-body1-170 MuiTypography-colorTextSecondary-194 MuiListItemText-secondary-159"
            >
              I need to be tested
            </p>
          </div>
          <span
            class="MuiTouchRipple-root-203"
          />
        </div>
        <div
          class="MuiListItemSecondaryAction-root-196"
        >
          <button
            aria-label="Delete"
            class="MuiButtonBase-root-142 MuiIconButton-root-197"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label-202"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-146"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M19 4H5c-1.11 0-2 .9-2 2v12c0 1.1.89 2 2 2h4v-2H5V8h14v10h-4v2h4c1.1 0 2-.9 2-2V6c0-1.1-.89-2-2-2zm-7 6l-4 4h3v6h2v-6h3l-4-4z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root-203"
            />
          </button>
        </div>
      </li>
      <hr
        class="MuiDivider-root-127"
      />
      <li
        class="MuiListItem-container-132"
      >
        <div
          class="MuiButtonBase-root-142 MuiListItem-root-131 MuiListItem-default-134 MuiListItem-gutters-138 MuiListItem-button-139 MuiListItem-secondaryAction-140 chatListItem"
          role="button"
          tabindex="0"
        >
          <div
            class="MuiListItemIcon-root-145"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-146"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </div>
          <div
            class="MuiListItemText-root-155"
          >
            <span
              class="MuiTypography-root-161 MuiTypography-subheading-168 MuiListItemText-primary-158"
            >
              Naru
            </span>
            <p
              class="MuiTypography-root-161 MuiTypography-body1-170 MuiTypography-colorTextSecondary-194 MuiListItemText-secondary-159"
            >
              I need to be mocked
            </p>
          </div>
          <span
            class="MuiTouchRipple-root-203"
          />
        </div>
        <div
          class="MuiListItemSecondaryAction-root-196"
        >
          <button
            aria-label="Delete"
            class="MuiButtonBase-root-142 MuiIconButton-root-197"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label-202"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-146"
                focusable="false"
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M0 0h24v24H0z"
                  fill="none"
                />
                <path
                  d="M19 4H5c-1.11 0-2 .9-2 2v12c0 1.1.89 2 2 2h4v-2H5V8h14v10h-4v2h4c1.1 0 2-.9 2-2V6c0-1.1-.89-2-2-2zm-7 6l-4 4h3v6h2v-6h3l-4-4z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root-203"
            />
          </button>
        </div>
      </li>
      <hr
        class="MuiDivider-root-127"
      />
    </nav>
    
  </div>
</div>
`;

exports[`@communication/messenger: should render the messenger empty state: Representing the empty state without chats. 1`] = `
<div>
  <div
    class="MuiPaper-root-1 MuiPaper-elevation2-5 MuiPaper-rounded-2 messenger"
  >
    <div
      class="messengerAppBar"
    >
      <header
        class="MuiPaper-root-1 MuiPaper-elevation4-7 MuiAppBar-root-28 MuiAppBar-positionFixed-29 MuiAppBar-colorPrimary-35 mui-fixed"
      >
        <div
          class="MuiToolbar-root-37 MuiToolbar-regular-39 MuiToolbar-gutters-38"
        >
          Messages
        </div>
      </header>
    </div>
    <div
      class="emptyState"
    >
      <div
        class="emptystateItems"
      >
        <h2>
          You actually have no active conversations.
        </h2>
        <button
          class="MuiButtonBase-root-67 MuiButton-root-41 MuiButton-contained-52 MuiButton-containedPrimary-53 MuiButton-raised-55 MuiButton-raisedPrimary-56"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label-42"
          >
            Find an advert
          </span>
          <span
            class="MuiTouchRipple-root-70"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
